File: Tutorials/Tuckshop/Client/Tuckshop.WebApp/src/Domain/Models/Customers/Commands/UpdateCustomerDetails.ts
import { ModelBase, Rules, Validation } from '@singularsystems/neo-core';
export default class UpdateCustomerDetails extends ModelBase {
    static typeName = "UpdateCustomerDetails";

    constructor() {
        super();
        this.makeObservable();
    }

    @Rules.Required()
    public customerId: number = 0;

    @Rules.Required()
    @Rules.StringLength(100)
    public firstName: string = "";

    @Rules.Required()
    @Rules.StringLength(100)
    public lastName: string = "";

    @Rules.Required()
    @Rules.StringLength(100)
    public email: string = "";

    @Rules.StringLength(100)
    public phoneNumber: string = "";

    // Client only properties / methods

   addBusinessRules(rules: Validation.Rules<UpdateCustomerDetails>) {
        super.addBusinessRules(rules);

        rules.failWhen(c => !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(c.email),
        "Please enter a valid email address."
    );

     rules.failWhen(
        c => !/^\d{10}$/.test(c.phoneNumber),
        "Phone number must be exactly 10 digits."
    );

    rules.failWhen(
    c => /^\d{10}$/.test(c.phoneNumber) && /^(\d)\1{9}$/.test(c.phoneNumber),
    "Please enter a valid phone number."
);
    }

    public toString(): string {
        if (this.isNew || !this.firstName) {
            return "New update customer details";
        } else {
            return this.firstName;
        }
    }
}

File: Tutorials/Tuckshop/Server/Tuckshop.Domain/Tuckshop.Core.Models/Models/Customers/Commands/UpdateCustomerDetails.cs
namespace Tuckshop.Core.Models.Customers.Commands
{
  using Neo.Model;
  using Neo.Model.Validation;
  using System.ComponentModel.DataAnnotations;

  /// <summary>
  /// Class representing an update to a customer's profile details.
  /// Deliberately does not include WalletBalance — this command is
  /// structurally incapable of touching it.
  /// </summary>
  public class UpdateCustomerDetails : ModelBase<UpdateCustomerDetails>
  {
    /// <summary>
    /// Gets and sets the Customer Id.
    /// </summary>
    [Required]
    public int CustomerId { get; set; }

    /// <summary>
    /// Gets and sets the customer's first name.
    /// </summary>
    [Required(AllowEmptyStrings = false)]
    [StringLength(100)]
    public string FirstName { get; set; } = string.Empty;

    /// <summary>
    /// Gets and sets the customer's last name.
    /// </summary>
    [Required(AllowEmptyStrings = false)]
    [StringLength(100)]
    public string LastName { get; set; } = string.Empty;

    /// <summary>
    /// Gets and sets the customer's email address.
    /// </summary>
    [Required(AllowEmptyStrings = false)]
    [StringLength(100)]
    public string Email { get; set; } = string.Empty;

    /// <summary>
    /// Gets and sets the customer's phone number.
    /// </summary>
    ///  [Required(AllowEmptyStrings = false)]
    [StringLength(100)]
    public string PhoneNumber { get; set; } = string.Empty;

    /// <inheritdoc />
    protected override void AddBusinessRules(ValidationRules<UpdateCustomerDetails> rules)
    {
      rules.FailWhen(
        c => !System.Text.RegularExpressions.Regex.IsMatch(c.PhoneNumber ?? "", @"^\d{10}$"),
        "Phone number must be exactly 10 digits.");

      rules.FailWhen(
    c => System.Text.RegularExpressions.Regex.IsMatch(c.PhoneNumber ?? "", @"^\d{10}$")
         && System.Text.RegularExpressions.Regex.IsMatch(c.PhoneNumber ?? "", @"^(\d)\1{9}$"),
    "Please enter a valid phone number.");

      rules.FailWhen(
        c => !System.Text.RegularExpressions.Regex.IsMatch(c.Email ?? "", @"^[^\s@]+@[^\s@]+\.[^\s@]+$"),
        "Please enter a valid email address");
    }
  }
}